NMS-13972: Document housekeeping tasks before upgrade (#4477)
OpenNMS/opennms: docs/modules/deployment/pages/upgrade/introduction.adoc

[[upgrade]]
= Upgrade {page-component-title}

This section outlines the steps to follow to upgrade to a newer version of {page-component-title}.
You may need to complete additional steps to upgrade in a more complex setup (running more than one OpenNMS instance, database migration requirements, age of current version, and so on).

A basic upgrade with one instance takes about an hour.
This time varies, and will be longer depending on how customized and complex your setup is.

[[byb_upgrade]]
== Before you begin

There are a number of tasks you should complete before starting an upgrade:

> * Check the https://vault.opennms.com/docs/opennms/releases/index.html[release notes] for breaking changes in each release between your current version and the version you want to upgrade to.

Reviewer: There is nothing in the vault starting with v28, and trying v28 on the docs site redirects to v29.  Would need to use github branches to get v28 (and v29 once 30 releases)

Author: Agreed. I'm hoping that we will have this issue addressed before 30 goes out, but I will add a note here about 28 in case we don't.

Address any issues that might affect you.
* Stop OpenNMS and PostgreSQL
* xref:operation:admin/housekeeping/introduction.adoc#prune-events[Prune unneeded events]
* <<non-root-permissions, Update permissions>> (if upgrading from pre-29 Horizon or pre-2022 Meridian)
* Back up your database (same day/state)
* Back up RRD files (same day/state)
* Back up the opennms /etc directory (opennms folder) (same day/state)
* <<events-table,Assess the events table size>> (especially important if you have a particularly large network)
* xref:deployment:upgrade/diff.adoc#run_diff[Identify the configuration files] you have changed to determine what you need to migrate as part of an upgrade.

[[non-root-permissions]]
== Update permissions (if upgrading from pre-29 versions)

ifeval::["{page-component-title}" == "Horizon"]
Starting with version 29, {page-component-title} runs as the `opennms` user rather than root by default.
If you are upgrading from a pre-29 version of {page-component-title}, you must fix the ownership of your files before upgrading.

> For information on how to do this, see https://docs.opennms.com/horizon/29/releasenotes/whatsnew.html?q=non-root#running-as-non-root[Running as non-root].

Reviewer: Remove search highlighting from link
```suggestion
For information on how to do this, see https://docs.opennms.com/horizon/29/releasenotes/whatsnew.html#running-as-non-root[Running as non-root].
```

endif::[]

ifeval::["{page-component-title}" == "Meridian"]
Starting with {page-component-title} 2022, {page-component-title} runs as the `opennms` user rather than root by default.
If you are upgrading from a pre-2022 {page-component-title} version, you must fix the ownership of your files before upgrading.

> For information on how to do this, see https://docs.opennms.com/meridian/2022/releasenotes/whatsnew.html[Running as non-root].

Reviewer: Change direct link to subsection anchor
```suggestion
For information on how to do this, see https://docs.opennms.com/meridian/2022/releasenotes/whatsnew.html#running-as-non-root[Running as non-root].
```

[[byb_upgrade]]
== Before you begin

There are a number of tasks you should complete before starting an upgrade:

* Check the release notes for breaking changes in each release between your current version and the version you want to upgrade to.
Address any issues that might affect you.
* Stop OpenNMS and PostgreSQL
* xref:operation:admin/housekeeping/introduction.adoc#prune-events[Prune unneeded events]
* xref:operation:admin/housekeeping/introduction.adoc#non-root-permissions[Update permissions] (if upgrading from pre-29 Horizon or pre-2022 Meridian)
* ##Back up your database## (same day/state)
* ##Back up RRD files## (same day/state)
* ##Back up the opennms /etc directory## (opennms folder) (same day/state)
* <<events-table,Assess the events table size>> (especially important if you have a particularly large network)
* xref:deployment:upgrade/diff.adoc#run_diff[Identify the configuration files] you have changed to determine what you need to migrate as part of an upgrade.

[[non-root-permissions]]
== Update permissions (if upgrading from pre-29 versions)

ifeval::["{page-component-title}" == "Horizon"]

> Since version 29, {page-component-title} runs as the opennms user rather than root by default.

Reviewer: ```suggestion
Starting with version 29, {page-component-title} runs as the `opennms` user rather than root by default.
```

If you are upgrading from a pre-29 version of {page-component-title}, you must fix the ownership of your files before upgrading.

For information on how to do this, see https://docs.opennms.com/horizon/29/releasenotes/whatsnew.html?q=non-root#running-as-non-root[Running as non-root].
endif::[]

ifeval::["{page-component-title}" == "Meridian"]

> Since {page-component-title} 2022, {page-component-title} runs as the opennms user rather than root by default.

Reviewer: ```suggestion
Starting with {page-component-title} 2022, {page-component-title} runs as the `opennms` user rather than root by default.
```

[[upgrade]]
= Upgrade {page-component-title}

This section outlines the steps to follow to upgrade to a newer version of {page-component-title}.
You may need to complete additional steps to upgrade in a more complex setup (running more than one OpenNMS instance, database migration requirements, age of current version, and so on).

> A basic upgrade with one instance takes about an hour.
> This time varies, and will be longer depending on how customized and complex your setup is.

Reviewer: Personally, I'd like to inject a little ambiguity into the amount of time.
```suggestion
A basic upgrade of one instance can take about an hour.
This time varies, and will be longer depending on how customized and complex your setup is.
```

[[byb_upgrade]]
== Before you begin

There are a number of tasks you should complete before starting an upgrade:

ifeval::["{page-component-title}" == "Meridian"]
* Check the https://meridian.opennms.com/[release notes] for breaking changes in each release between your current version and the version you want to upgrade to.
Address any issues that might affect you.
endif::[]

ifeval::["{page-component-title}" == "Horizon"]
* Check the https://vault.opennms.com/docs/opennms/releases/index.html[release notes] for breaking changes in each release between your current version and the version you want to upgrade to.
Address any issues that might affect you.

NOTE: Release notes for versions 28 and 29 are not available at the vault docs link (above) at this time.

> Please check the branches on GitHub for those release notes.

Reviewer: ```suggestion
Please check the branches on https://github.com/OpenNMS/opennms/releases[GitHub] for those release notes.
```

endif::[]

* Stop OpenNMS and PostgreSQL
* xref:operation:admin/housekeeping/introduction.adoc#prune-events[Prune unneeded events]

> * <<non-root-permissions, Update permissions>> (if upgrading from pre-29 Horizon or pre-2022 Meridian)

Reviewer: This step should be done closer to the end of the upgrade process, after installing updated packages and reconciling config file changes but before starting services. The script to fix permissions isn't available until after 29/2022 packages are installed.

Author: Removed, and will add as part of upgrade process procedure when that is added.

* Back up your database (same day/state)
* Back up RRD files (same day/state)
* Back up the opennms /etc directory (opennms folder) (same day/state)
* <<events-table,Assess the events table size>> (especially important if you have a particularly large network)
* xref:deployment:upgrade/diff.adoc#run_diff[Identify the configuration files] you have changed to determine what you need to migrate as part of an upgrade.

[[non-root-permissions]]
== Update permissions (if upgrading from pre-29 versions)

ifeval::["{page-component-title}" == "Horizon"]
Starting with version 29, {page-component-title} runs as the `opennms` user rather than root by default.

> If you are upgrading from a pre-29 version of {page-component-title}, you must fix the ownership of your files before upgrading.

Reviewer: ```suggestion
If you are upgrading from a pre-29 version of {page-component-title}, you must fix the ownership of your files as part of the upgrade process.
```
